tests/arns.test.mjs

```diff
@@ -1509,4 +1513,226 @@ describe('ArNS', async () => {
+          type: 'lease',
+        });
+
+        const balancesAfter = await getBalances({
+          memory: extendResult.Memory,
+        });
+        const expectedCost = 100_000_000 * 0.8;
+        assert.equal(
+          balancesAfter[joinedGateway],
+          balancesBefore[joinedGateway] - expectedCost,
+        );
+      });
+
+      it('upgrading leases to permabuys should include ArNS discount', async () => {
+        const balancesBefore = await getBalances({
+          memory: buyRecordResult.Memory,
+        });
+        const upgradeResult = await handle(
+          {
+            From: joinedGateway,
+            Owner: joinedGateway,
+            Tags: [
+              { name: 'Action', value: 'Upgrade-Name' },
+              { name: 'Name', value: 'great-name' },
+            ],
+            Timestamp: afterDistributionTimestamp,
+          },
+          buyRecordResult.Memory,
+        );
+
+        const record = JSON.parse(upgradeResult.Messages[0].Data);
+        assert.deepEqual(record, {
+          processId: ''.padEnd(43, 'a'),
+          purchasePrice: basePermabuyPrice * 0.8,
+          startTimestamp: afterDistributionTimestamp,
+          name: 'great-name',
+          undernameLimit: 10,
+          type: 'permabuy',
+        });
+
+        const balancesAfter = await getBalances({
+          memory: upgradeResult.Memory,
+        });
+        const expectedCost = basePermabuyPrice * 0.8;
+        assert.equal(
+          balancesAfter[joinedGateway],
+          balancesBefore[joinedGateway] - expectedCost,
+        );
+      });
+
+      it('increasing undername limit should include ArNS discount', async () => {
+        const balancesBefore = await getBalances({
+          memory: buyRecordResult.Memory,
+        });
+        const increaseLimitResult = await handle(
+          {
+            From: joinedGateway,
+            Owner: joinedGateway,
+            Tags: [
+              { name: 'Action', value: 'Increase-Undername-Limit' },
+              { name: 'Name', value: 'great-name' },
+              { name: 'Quantity', value: '20' },
+            ],
+            Timestamp: afterDistributionTimestamp,
+          },
+          buyRecordResult.Memory,
+        );
+
+        const record = JSON.parse(increaseLimitResult.Messages[0].Data);
+        assert.deepEqual(record, {
+          processId: ''.padEnd(43, 'a'),
+          purchasePrice: baseLeasePrice * 0.8,
+          endTimestamp: 1751524800000,
+          startTimestamp: afterDistributionTimestamp,
+          undernameLimit: 30,
+          type: 'lease',
+        });
+
+        const balancesAfter = await getBalances({
+          memory: increaseLimitResult.Memory,
+        });
+        const priceOfOneUnderName = 500_000;
+        const expectedCost = priceOfOneUnderName * 20 * 0.8;
+        assert.equal(
+          balancesAfter[joinedGateway],
+          balancesBefore[joinedGateway] - expectedCost,
+        );
+      });
+    });
+  });
```

Comment:
TODO: Add a test for submitAuctionBid case. Just call anything like "getBalances" after the grace period is over (which is weeks PAST the endTimestamp of the record) and the prepend handler will cause an auction to get created. Then test pricing out the auction bid.

Reply:
8c35a96eac5e1df8dd966fe499b1d04af0e9d89e